src/hal/src/pso/mod.rs

```diff
@@ -221,32 +248,112 @@ pub struct SpecializationConstant {
 }
 
 /// Specialization information structure.
-#[derive(Debug, Copy)]
-pub struct Specialization<'a> {
+#[derive(Debug, Copy, Clone)]
+pub struct RawSpecialization<'a> {
     /// Constant array.
     pub constants: &'a [SpecializationConstant],
     /// Raw data.
     pub data: &'a [u8],
 }
 
-impl<'a> Default for Specialization<'a> {
+impl RawSpecialization<'_> {
+    /// Create empty specialization instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Default for RawSpecialization<'_> {
     fn default() -> Self {
-        Specialization {
+        RawSpecialization {
             constants: &[],
             data: &[],
         }
     }
 }
 
-impl<'a> Clone for Specialization<'a> {
-    fn clone(&self) -> Self {
-        Specialization {
-            constants: self.constants,
-            data: self.data,
+/// Convert typed specialization into untyped one.
+pub trait Specialization {
+    /// Build `Specialization` value.
+    fn as_raw(&self) -> RawSpecialization<'_>;
+}
+
+/// Temporary storage for specialization constants.
+#[derive(Clone, Debug, Default)]
+pub struct SpecializationStorage {
+    /// Actual data for specialization constants.
+    pub data: Vec<u8>,
```

Comment:
do we need to expose everything here to the user? I thought this is built with helper methods anyway

Reply:
What if user would want to make his own `Specialization` trait implementation?
This type get merged with original `Specialization` anyway though.